internal/history/models.go

```diff
@@ -0,0 +1,583 @@
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package history
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/stacklok/minder/internal/db"
+)
+
+var (
+	// ErrMalformedCursor represents errors in the cursor payload.
+	ErrMalformedCursor = errors.New("malformed cursor")
+	// ErrInvalidTimeRange is returned the time range from-to is
+	// either missing one end or from is greater than to.
+	ErrInvalidTimeRange = errors.New("invalid time range")
+	// ErrInclusionExclusion is returned when both inclusion and
+	// exclusion filters are specified on the same field.
+	ErrInclusionExclusion = errors.New("inclusion and exclusion filter specified")
+	// ErrInvalidIdentifier is returned when an identifier
+	// (e.g. entity name) is empty or malformed.
+	ErrInvalidIdentifier = errors.New("invalid identifier")
+)
+
+// Direction enumerates the direction of the Cursor.
+type Direction string
+
+const (
+	// Next represents the next page.
+	Next = Direction("next")
+	// Prev represents the prev page.
+	Prev = Direction("prev")
+)
+
+// ListEvaluationCursor is a struct representing a cursor in the
+// dataset of historical evaluations.
+type ListEvaluationCursor struct {
+	Time      time.Time
+	Direction Direction
+}
+
+var (
+	// DefaultCursor is a cursor starting from the beginning of
+	// the data set.
+	DefaultCursor = ListEvaluationCursor{
+		Time:      time.UnixMicro(999999999999999999).UTC(),
+		Direction: Next,
+	}
+)
+
+// ParseListEvaluationCursor interprets an opaque payload and returns
+// a ListEvaluationCursor. The opaque paylaod is expected to be of one
+// of the following forms
+//
+//   - `"+1257894000000000"` meaning the next page
+//     of data starting from the given UUID excluded
+//
+//   - `"-1257894000000000"` meaning the previous
+//     page of data starting from the given UUID excluded
+//
+//   - `"1257894000000000"` meaning the next page
+//     of data (default) starting from the given UUID excluded
+func ParseListEvaluationCursor(payload string) (*ListEvaluationCursor, error) {
+	decoded, err := base64.StdEncoding.DecodeString(payload)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrMalformedCursor, err)
+	}
+
+	switch {
```

Comment:
do we expect more cases in this switch-case? It seems like `+`, `-` and `default` only differ in the direction and whether they parse the first character or not..

Reply:
No we don't, and yes, you're correct they only differ in direction and first character.
Let me rewrite it a bit better.